http3/body.go

```diff
@@ -0,0 +1,68 @@
+package http3
+
+import (
+	"errors"
+	"io"
+)
+
+// The body of a http.Request or http.Response.
+type body struct {
+	str io.ReadCloser
+
+	isRequest bool
+
+	bytesRemainingInFrame uint64
+}
+
+var _ io.ReadCloser = &body{}
+
+func newRequestBody(str io.ReadCloser) *body {
+	return &body{
+		str:       str,
+		isRequest: true,
+	}
+}
+
+func newResponseBody(str io.ReadCloser) *body {
+	return &body{str: str}
+}
+
+func (r *body) Read(b []byte) (int, error) {
+	if r.bytesRemainingInFrame == 0 {
+	parseLoop:
+		for {
+			frame, err := parseNextFrame(r.str)
+			if err != nil {
+				return 0, err
+			}
+			switch f := frame.(type) {
+			case *headersFrame:
+				// skip HEADERS frames
+				continue
+			case *dataFrame:
+				r.bytesRemainingInFrame = f.Length
+				break parseLoop
+			default:
+				return 0, errors.New("unexpected frame")
+			}
+		}
+	}
+
+	var n int
+	var err error
+	if r.bytesRemainingInFrame < uint64(len(b)) {
+		n, err = r.str.Read(b[:r.bytesRemainingInFrame])
+	} else {
+		n, err = r.str.Read(b)
+	}
+	r.bytesRemainingInFrame -= uint64(n)
```

Comment:
Is there a risk in doing this before we check the error value?

Right now we don't have any temporary errors I believe, but if we added those in the future there could be an underflow here I imagine?

Reply:
`io.Reader` says:
>  It returns the number of bytes read (0 <= n <= len(p)) and any error encountered.

So I believe this should be safe with any `io.Reader`.